QuestionService transition

Hamid-Keshavarz opened the pull request.

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Controllers/QuestionController.cs:
```diff
@@ -26,11 +26,11 @@ public async Task<IActionResult> GetQuestions(int courseId, string language, str
 
         [HttpGet("unanswered")]
-        public async Task<IActionResult> GetUnansweredQuestions([FromQuery] QuestionParameter request)
+        public async Task<IActionResult> GetUnansweredQuestions(int courseId, string language, string system, string userId)
```
userId als erster Parameter und nullable

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Controllers/QuestionController.cs:
```diff
@@ -44,11 +44,11 @@ public async Task<IActionResult> GetUnansweredQuestions([FromQuery] QuestionPara
 
         [HttpGet("incorrect")]
-        public async Task<IActionResult> GetIncorrectQuestions([FromQuery] QuestionParameter request)
+        public async Task<IActionResult> GetIncorrectQuestions(int courseId, string language, string system, string userId)
```
userId als erster Parameter und nullable

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Controllers/QuestionController.cs:
```diff
@@ -26,11 +26,11 @@ public async Task<IActionResult> GetQuestions(int courseId, string language, str
             try
             {
-                var unansweredQuestions = await _questionService.GetUnansweredQuestionsAsync(request);
+                var unansweredQuestions = await _questionService.GetUnansweredQuestionsAsync(courseId, language, system, userId);
```
userId als erster Parameter und nullable

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Controllers/QuestionController.cs:
```diff
@@ -44,11 +44,11 @@ public async Task<IActionResult> GetUnansweredQuestions([FromQuery] QuestionPara
             try
             {
-                var incorrectQuestions = await _questionService.GetIncorrectQuestionsAsync(request);
+                var incorrectQuestions = await _questionService.GetIncorrectQuestionsAsync(courseId, language, system, userId);
```
userId als erster Parameter und nullable

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Controllers/QuestionController.cs:
```diff
@@ -63,11 +63,11 @@ public async Task<IActionResult> GetIncorrectQuestions([FromQuery] QuestionParam
 
         [HttpGet("correct")]
-        public async Task<IActionResult> GetCorrectQuestions([FromQuery] QuestionParameter request)
+        public async Task<IActionResult> GetCorrectQuestions(int courseId, string language, string system, string userId)
```
userId als erster Parameter und nullable

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Controllers/QuestionController.cs:
```diff
@@ -63,11 +63,11 @@ public async Task<IActionResult> GetIncorrectQuestions([FromQuery] QuestionParam
             try
             {
-                var correctQuestions = await _questionService.GetCorrectQuestionsAsync(request);
+                var correctQuestions = await _questionService.GetCorrectQuestionsAsync(courseId, language, system, userId);
```
userId als erster Parameter und nullable

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Controllers/QuestionController.cs:
```diff
@@ -82,22 +82,28 @@ public async Task<IActionResult> GetCorrectQuestions([FromQuery] QuestionParamet
 
         [HttpPost("updateAnswer")]
-        public async Task<IActionResult> UpdateQuestionStatus(int questionId, string questionStatus)
+        public async Task<IActionResult> UpdateQuestionStatus(int questionId, string userId, string questionStatus)
```
userId als erster Parameter und nullable

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Controllers/QuestionController.cs:
```diff
@@ -82,22 +82,28 @@ public async Task<IActionResult> GetCorrectQuestions([FromQuery] QuestionParamet
             {
-               await _questionService.UpdateQuestionStatusAsync(questionId, questionStatus);
-               return Ok(new { Message = "Fragestatus erfolgreich aktualisiert", UpdatedQuestionStatus = questionStatus });
+                await _questionService.UpdateQuestionStatusAsync(questionId, userId, questionStatus);
```
userId als erster Parameter und nullable

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Services/Interfaces/IQuestionService.cs:
```diff
@@ -5,11 +5,23 @@ namespace ShortcutTrainerBackend.Services.Interfaces
-        Task<IEnumerable<DtoQuestion>> GetCorrectQuestionsAsync(QuestionParameter request);
-        Task UpdateQuestionStatusAsync(int questionId, string questionStatus);
+
+        // GetUnansweredQuestions
```
userId immer als erstes und nullable. Alle Methoden fragen Fragen und Antworten in irgendeiner Form ab. Du kannst dir überlegen, welchen Teil der LINQ-Query alle gemeinsam haben und diese Abfrage in eine eigene private Methode packen. Damit erhöhst du die Lesbarkeit und reduzierst die LOC

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Data/Models/Answer.cs:
```diff
@@ -1,4 +1,5 @@
 using DevExpress.Xpo;
+using ShortcutTrainerBackend.Data.TransferObjects;
```
Ist obsolet. Außerdem kann QuestionParameter gelöscht werden

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Services/QuestionService.cs:
```diff
@@ -33,30 +33,154 @@ public IEnumerable<DtoQuestion> GetQuestions(int courseId, string language, stri
         }
-        
+
         public async Task<IEnumerable<DtoQuestion>> GetQuestionsAsync(int courseId, string language, string system)
```
GetQuestions könnte auch erweitert werden und alle Fragen und Antworten samt Status liefern. Dafür auch wieder userId als erster Parameter und wieder nullable

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Services/QuestionService.cs:
```diff
@@ -33,30 +33,154 @@ public IEnumerable<DtoQuestion> GetQuestions(int courseId, string language, stri
+        public IEnumerable<DtoQuestion> GetUnansweredQuestions(int courseId, string language, string system, string userId)
+        {
+            var questions = new XPCollection<Question>(_session)
+                .Where(q => q.Course.Id == courseId && q.Course.Language == language)
```
Bei gegebenem User, also userId != null, kannst du auch über UserCourse abfragen

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Services/QuestionService.cs:
```diff
@@ -33,30 +33,154 @@ public IEnumerable<DtoQuestion> GetQuestions(int courseId, string language, stri
+                            Shortcut = a.Shortcut
+                        })
+                }).Where(q =>
+            q.Answers.Any() &&
```
Die Einrückung beachten -> mit jedem Sub-Command ein Tab weiter rein

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Services/QuestionService.cs:
```diff
@@ -33,30 +33,154 @@ public IEnumerable<DtoQuestion> GetQuestions(int courseId, string language, stri
+                                userAnswer.QuestionStatus = questionStatus;
+
+                                // Commit der Transaktion
+                                await Task.Run(() => unitOfWork.CommitChanges());
```
Du benötigst kein UnitOfWork.CommitChanges -> das du nur ein Objekt speichern möchtest, kannst du auch
await _session.SaveChangesAsync(<Dein zu speicherndes Objekt>); verwenden

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Services/QuestionService.cs:
```diff
@@ -33,30 +33,154 @@ public IEnumerable<DtoQuestion> GetQuestions(int courseId, string language, stri
         {
-            throw new NotImplementedException();
+            try
+            {
```
Das Statement ist insgesamt zu komplex. Du brauchst kein 3-fach-verschachteltes If-Statement. Außerdem existieren die Funktionen, die du verwendest auch schon für die Klasse Session. Ohne das Statement noch tiefergehender zu überprüfen, überlege dir am besten was du brauchst und wie du es am geschicktesten Aufbauen kannst

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Controllers/CourseController.cs:
```diff
@@ -18,9 +18,9 @@ public CourseController(ILogger<CourseController> logger, ICourseService courses
+        public async Task<IActionResult> GetCourses(string language, string? system, string? searchString, int? limit)
         {
-            return Ok(await _coursesService.GetCoursesAsync(userId, language, system, searchString, limit));
+            return Ok(await _coursesService.GetCoursesAsync(language, system, searchString, limit));
```
warum hast du eine Änderung, die die Course-Route betreffen?
Im CourseService und im CourseController sind Fehler. Daher warten bis die Kurs-Überarbeitung gemerged ist und dann einfach aktualisieren

Hamid-Keshavarz replied:
Ich habe etwas ausprobiert und einfach vergessen, die Änderung rückgängig zu machen.

BravoEchoNov commented on ShortcutTrainerBackend/ShortcutTrainerBackend/Services/QuestionService.cs:
```diff
@@ -14,49 +14,132 @@ public QuestionService(Session session)
+        }
+
+        public IEnumerable<DtoQuestion> GetUnansweredQuestions(string? userId, int courseId, string language, string system)
+        {
```
es fehlt noch ein Mechanismus, der unbeantwortete Fragen in der DB ergänzt, da es bisher noch nichts gibt. Es wäre unter Umständen gar nicht notwendig einen separaten Status "Unanswered" zu haben. D.h. du könntest auch einfach alle Fragen zurückgeben, für die es für die angegebenen Parameter noch keine Einträge gibt (also abhängig von Language und System)

Hamid-Keshavarz committed:
Question feature (#34)

* QuestionController has been extended
* IQuestionService has been extended
* The following methods has been addes to the QuestionService:
	GetQuestionsByStatus
	GetQuestions
	GetQuestionsAsync
	GetUnansweredQuestions
	GetUnansweredQuestionsAsync
	GetIncorrectQuestions
	GetIncorrectQuestionsAsync
	GetCorrectQuestions
	GetCorrectQuestionsAsync
	UpdateQuestionStatusAsync

Hamid-Keshavarz committed:
Removal of 'using ShortcutTrainerBackend.Data.Models;'
